### DISCUSSION TITLE
how should I architect a project that maps between regex syntaxes? (should I target `Ast` or `Hir` in `regex-syntax`?)

### DISCUSSION THREAD

**Discussion by cosmicexplorer** [rust-lang/regex#1167] [Q&A] [answered]
As per this issue comment (https://github.com/rust-lang/regex/issues/962#issuecomment-1963008400):
I was looking through this repo to see whether anyone had translated from [emacs regexp syntax](https://www.gnu.org/software/emacs/manual/html_node/emacs/Regexps.html) using regex-syntax yet (my eventual goal is to produce an emacs native module). I'm pretty sure I could do that by transpiling the emacs regexp pattern string to a [`regex_syntax::ast::Ast`](https://docs.rs/regex-syntax/latest/regex_syntax/ast/index.html), since the AST looks pretty simple, and emacs regexps don't have any special features: it pushes that to the language instead.

As mentioned by @db48x in https://github.com/rust-lang/regex/issues/962#issuecomment-1446384219, emacs provides a high-level `rx` sexp notation, but I think that actually aligns with @BurntSushi's framing of "using your programming language features to assemble the regex" from https://github.com/rust-lang/regex/issues/962#issuecomment-1446263381! So I was hoping to demonstrate that regex-syntax makes that sort of thing possible *without* implementing that sort of higher-level abstraction itself directly (this is actually what makes it more generic/reusable)!

At first glance I think e.g. `emacs-regexp-syntax` can just be its own crate that depends on `regex-syntax` and generates a `regex_syntax::ast::Ast`, and I can propose upstreaming any extensions I need from the generic regex-syntax AST to this repo (although this seems unlikely to be necessary). I can then have another crate `emacs-regexp` which depends on `regex-automata` and exposes the emacs native module interface.

**Comment by BurntSushi** [comment, marked as answer]
Aye, just copying my answer here from #962 so that this can be marked as answered. :-) Feel free to edit your comment to remove the other copy hah.

-----

> @BurntSushi is that how you would suggest architecting this if I wanted to interface other regexp interfaces to `regex-syntax` and co?

Maybe? I've never done the exercise before, so I'm not sure. It's _plausible_ the right target is `regex-syntax::hir::Hir`. Namely, an `Hir` is perhaps closer to what you might think of as a traditional AST where as an `Ast` is perhaps closer to a [Concrete Syntax Tree](https://eli.thegreenplace.net/2009/02/16/abstract-vs-concrete-syntax-trees/). A concrete syntax tree essentially tries to describe the concrete syntax _precisely_ so that it can be round-tripped. This makes it very detailed and possibly contains a lot of stuff you don't need to care about. (Regrettably, it isn't a true concrete syntax tree because of comments and whitespace in `x` mode. But it's close.)

Either target is reasonable IMO. With that said, if you need to deal with Unicode, then targeting the `Ast` potentially gives you some very big benefits. The `Ast -> Hir` conversion is what resolves things like `[\p{Letter}&&\p{Greek}]` into a single sequence of non-overlapping non-adjacent ranges of codepoints. That is a surprisingly non-trivial thing to do. If you only care about ASCII though, then you could feasibly target `Hir` without too much work.

**Reply by cosmicexplorer** [reply]
Since I *definitely* care about Unicode, I will take this as an argument to try for the `Ast` conversion! I am currently worried much less about the AST/CST distinction than whether the `Ast` structure is somehow incompatible (not round-trippable) with the emacs pattern syntax. However, I suspect any such dissonance can probably be resolved pretty cleanly.

For example, [emacs has `\_<` and `\_>` for symbol boundaries](https://www.gnu.org/software/emacs/manual/html_node/emacs/Regexp-Backslash.html), which are like word boundaries, but intended to match symbols in programming languages (often used for syntax highlighting). Emphasis mine: 
> `\_<` matches the empty string, but only at the beginning of a symbol. A symbol is a sequence of one or more symbol-constituent characters. A symbol-constituent character is a character whose syntax is either ‘w’ or ‘_’. ‘\_<’ matches at the beginning of the buffer only if a symbol-constituent character follows. **As with words, the syntax table determines which characters are symbol-constituent.**

My first thought was to add something like `EmacsSymbolBoundary` as a case to [`regex_syntax::ast::AssertionKind`](https://docs.rs/regex-syntax/latest/regex_syntax/ast/enum.AssertionKind.html). However, after considering the emphasized sentence, I realized that emacs's ability to dynamically rebind word/symbol characters would probably require an enum case for `AssertionKind` that can reference some parameterized input, and no current `AssertionKind` case has any fields! So then I thought it would probably be better to propose a new `ParameterizedAssertionKind` node which generates HIR for "boundary, but match my particular definition of word" according to the input field (e.g. a `Vec<char>`). I am hoping that can be done without a lot of tricky special code and achieve similar performance to the existing ASCII/Unicode word class boundary processing.

I will get more familiar with the process of HIR generation for character classes (as well as the particulars of emacs regexp syntax) and see how to minimize the number of changes needed.

**Reply by BurntSushi** [reply]
Hmmm okay, I perhaps misunderstood your question.

My tolerance for adding things to `regex-syntax` that aren't actually used or supported by `regex` is pretty low. I don't really want it to become a kitchen sink for a bunch of different flavors. And indeed, adding new kinds of assertions that aren't supported by `regex` is not really something I want to pursue/maintain. And just as an example, adding something like `ParameterizedAssertionKind` and actually having it supported all the way through `regex-automata` is a monumental change. 

The _primary_ purpose of `regex-syntax` is to be an [implementation detail of `regex`](https://github.com/rust-lang/regex/tree/master/regex-syntax#motivation). Having its scope increased to support different kinds of syntaxes isn't really something I'm capable of doing unfortunately.

**Reply by BurntSushi** [reply]
And in the general case, it is very difficult to map one regex's features to another. I'm not familiar with emacs regexes, but they might for example do leftmost-longest matching and not leftmost-first. That's not even an artifact of the syntax at all. It's deeply embedded in the regex engine match semantics itself.

Basically, the only real way to use emacs regex from Rust is to write a binding or completely re-implement emacs regex to match it precisely. Any other avenue is going to require some kind of compromise that sounds like, "in emacs regex `foo` does `A`, but in this context, it does `B` or is unsupported entirely." So under that regime, you could say things like "`\_<` is unsupported."

**Reply by cosmicexplorer** [reply]
> And indeed, adding new kinds of assertions that aren't supported by regex is not really something I want to pursue/maintain.

> The primary purpose of regex-syntax is to be an [implementation detail of regex](https://github.com/rust-lang/regex/tree/master/regex-syntax#motivation). Having its scope increased to support different kinds of syntaxes isn't really something I'm capable of doing unfortunately.

Thank you, this is actually exactly the kind of feedback I was looking for by asking this question!

> adding something like `ParameterizedAssertionKind` and actually having it supported all the way through `regex-automata`

Hmmm, I was really hoping to only change the AST and not introduce any new HIR! But after seeing that [the HIR statically encodes the types of look-around assertions](https://docs.rs/regex-syntax/latest/regex_syntax/hir/enum.Look.html#variant.WordStartAscii) (which btw is a perfectly reasonable thing to do), I can absolutely see why that approach would require much more effort than I expected!

> Basically, the only real way to use emacs regex from Rust is to write a binding or completely re-implement emacs regex to match it precisely.

Hmmm, so we may have traded misinterpretations! I was absolutely not planning to invoke the emacs regex engine, but rather to replace it entirely with a Rust engine implemented via `regex-automata`! If successful, I was going to propose its inclusion within emacs itself. I agree that the semantics of how emacs invokes searches will be arcane and idiosyncratic, but I was actually hoping to use that as a sort of stress test for `regex-automata` as a mechanism for building other regex engines.

However, given your feedback above, it sounds like `regex-automata` also won't be able to handle a lot of extra weight attached for stuff that's not getting directly used in `regex`, so I'll also try to structure any extensions to `regex-syntax` or `regex-automata` as separate libraries. I know you've also split out crates like `memchr` already to avoid a dependency on the whole `regex` ecosystem to get the benefits of your great SIMD work, so it's also possible for me to avoid that entirely.

I was actually looking for an excuse to write another regex engine anyway so I have absolutely no problem with your very reasonable desire to keep `regex-{syntax,automata}` focused on serving the needs of `regex`! But I will definitely be looking for more ways to share code :)

**Reply by BurntSushi** [reply]
> Hmmm, I was really hoping to only change the AST and not introduce any new HIR! But after seeing that [the HIR statically encodes the types of look-around assertions](https://docs.rs/regex-syntax/latest/regex_syntax/hir/enum.Look.html#variant.WordStartAscii) (which btw is a perfectly reasonable thing to do), I can absolutely see why that approach is much more difficult than I expected!

It's even statically encoded [in `regex-automata` too](https://github.com/rust-lang/regex/blob/d7f9347f2a8a7f4e7583c88876411da12a09b572/regex-automata/src/util/look.rs#L25-L135). 

Then there's nastiness for computing all of the different starting configurations of a DFA: https://github.com/rust-lang/regex/blob/d7f9347f2a8a7f4e7583c88876411da12a09b572/regex-automata/src/util/start.rs#L193-L250

And then picking the right starting configuration when beginning a search: https://github.com/rust-lang/regex/blob/d7f9347f2a8a7f4e7583c88876411da12a09b572/regex-automata/src/util/determinize/mod.rs#L581-L682

> Hmmm, so we may have traded misinterpretations! I was absolutely not planning to invoke the emacs regex engine, but rather to replace it entirely with a Rust engine implemented via regex-automata!

Right. What I'm saying is that if you want a drop-in replacement for Emacs' regex engine, you'll _likely_ either need to completely roll your own implementation that matches it precisely or give up on a pure Rust solution and use a binding of some kind. I use the weasel word "likely" here on the off chance that there is some way to map its semantics in a precisely compatible way to `regex`. The existence of `\_<` is an _obvious_ example of a difference, but I'd bet my bottom dollar that there are much subtler differences.

The right prior here IMO is this: _unless a regex engine is implemented to strictly conform to an existing spec, then the implementation provides its own distinct dialect of regex syntax and semantics_. The only regex specs I know of in broad use are POSIX and ECMAScript.

> but I was actually hoping to use that as a sort of stress test for regex-automata as a mechanism for building other regex engines.

`regex-automata` is not really a framework for building arbitrary regex engines unfortunately. Its first and primary purpose is an implementation detail of `regex`. Its secondary purpose is to expose `regex` internals that permits composition _within the bounds of what the `regex` crate does._ A true framework for supporting arbitrary regex engines is really a completely different beast entirely, and would, IMO, probably be an _order of magnitude_ more work, to a first approximation. I don't even know if it could be done. The API design alone feels intractable.

> However, given your feedback above, it sounds like regex-automata also won't be able to handle a lot of extra weight attached for stuff that's not getting directly used in regex, so I'll also try to structure any extensions to regex-syntax or regex-automata as separate libraries. I know you've also split out crates like memchr already to avoid a dependency on the whole regex ecosystem to get the benefits of your great SIMD work, so it's also possible for me to avoid that entirely.
>
> I was actually looking for an excuse to write another regex engine anyway so I have absolutely no problem with your very reasonable desire to keep regex-{syntax,automata} focused on serving the needs of regex! But I will definitely be looking for more ways to share code :)

Aye yeah. You'll likely need your own crates. I would definitely encourage you to copy & paste liberally from `regex-syntax` and `regex-automata`, as needed. There are likely many things in there that could be used independently of anything else (with perhaps some modifications). Basically, the `regex` crate dependency tree could actually be _many_ more crates than it is. But I try really hard to balance that against the development and maintenance costs of multiple crates. That is, adding another required crate to the `regex` dependency tree is something that requires a very compelling motivation from my perspective.

But yes, it is very plausible that `memchr` and `aho-corasick` will be usable as-is for you. Although they are likely only useful as optimizations. Notably, they are entirely optional dependencies of `regex`.

Finally, I might encourage you to look at `regex-lite`. It's ~4,000 lines that is basically, "`regex`, but without Unicode and slower." It might be a more digestible starting point for you.

**Reply by cosmicexplorer** [reply]
> I use the weasel word "likely" here on the off chance that there is some way to map its semantics in a precisely compatible way to `regex`. The existence of `\_<` is an obvious example of a difference, but I'd bet my bottom dollar that there are much subtler differences.

On this point, I especially suspect the way that emacs regexp search is designed with respect to a stateful editor API means these subtle differences may show up across different types of inputs with different mutability guarantees (process output, temp buffer, narrowed/indirect buffer). It's very powerful to mix language, editor, and regex engine like this, but it's also why I was ready for the semantics to potentially be more difficult than other engines to mimic.

> A true framework for supporting arbitrary regex engines is really a completely different beast entirely, and would, IMO, probably be an order of magnitude more work, to a first approximation. I don't even know if it could be done. The API design alone feels intractable.

So I first wanted to solve this by working really hard on [interop between regex engines via improving packaging](https://github.com/spack/spack/issues/41960)! But that wouldn't address the issue that regex engines really are used in so many different idiosyncratic ways that I agree a truly general API design for it feels intractable! Also, regex aren't always the right solution to every problem and making a *truly* general regex engine sounds in the limit just as difficult as making a *truly* general parsing framework!

Instead, I am currently brainstorming a general "parser composition" framework which can hopefully be used to efficiently dispatch to black-box parsers like regex engines or literal matchers for parsing subsections of a more complex pattern. Ideally, this would mean that I could write my own parsing logic with any stackful/context-sensitive/dynamic stuff, and delegate to `regex`, `re2`, etc for the parts of my grammar which have standard semantics and are easy to optimize, and thereby get the benefits of accelerated automatons/SIMD matching where possible without forcing the regex engine to support every single type of pattern/precondition itself.

I have a much more specific use case in mind at first (I have a general parsing method which would be able to use this API), but that's one way I'm currently thinking of architecting this sort of thing.

> Basically, the regex crate dependency tree could actually be many more crates than it is. But I try really hard to balance that against the development and maintenance costs of multiple crates. That is, adding another required crate to the regex dependency tree is something that requires a very compelling motivation from my perspective.

Thanks, this is a super reasonable policy and it's helpful to say it out loud! 

> Finally, I might encourage you to look at regex-lite. It's ~4,000 lines that is basically, "regex, but without Unicode and slower." It might be a more digestible starting point for you.

Thanks so much!! I'm planning to build some benchmark harnesses to research the above composition idea and tiny regex impls are exactly what I would want to base off of!

**Reply by cosmicexplorer** [reply]
@junyer just pointed me to his analysis of an old proposal to adapt re2 to emacs regexps (https://github.com/google/re2/issues/127#issuecomment-267935881), which helpfully describes changes needed to support searching across emacs gap buffers (non-contiguous). I know non-contiguous/stream searching has long been high on your list of "would be awesome to have but would be a ton of work and hard to maintain", so this is another reason emacs semantics especially might be more difficult to support with `regex-{syntax,automata}` than other projects that are under 40 years old.

My parser composition framework *would* support non-contiguous input data in general, but it wouldn't be able to magically parse anything itself: the black-box parsers I would rely on would individually need to be able to support non-contiguous data. So that's another bit of functionality we might be able to eventually collaborate on and/or decouple into a shared crate (although that would probably be quite a large project to be clear, just another example is all).

**Reply by BurntSushi** [reply]
> and delegate to regex, re2, etc for the parts of my grammar which have standard semantics and are easy to optimize, and thereby get the benefits of accelerated automatons/SIMD matching where possible without forcing the regex engine to support every single type of pattern/precondition itself.

One thing potentially worth mentioning here is that the SIMD optimizations in the `regex` crate (including the DFA's "accelerated" states) are primarily _throughput_ optimizations. They aren't things that are likely to help with "parsing" tasks. They _can_, especially if there's a fast path for rejecting a match. But in general, I'd expect such optimizations to either be ineffectual or even lead to poorer performance on parsing tasks than if they didn't exist.

To be clear, I don't mean to say that this is a property of all SIMD code. But given how its used in `regex` and its overall architecture, they don't really help much with parsing. The SIMD is all about quickly finding the _next candidate start position of a match_. If the next candidate is always the next byte or a couple bytes away (as might be the case in a standard parsing task), then it doesn't really matter how fast you find that candidate.

Generally speaking, I think the kind of composition you're looking for might not work if performance is a goal. In my experience, performance and composition don't always go well together. Indeed, that's a big reason why `regex-automata` was so hard to build. Especially its API. The way I got around perf and composition not combining well is by _coupling_. For example, you can't use the lazy DFA in `regex-automata` with anything other than a `regex-automata::nfa::thompson::NFA`. They are coupled, but compositional, components. Introducing composition at an even higher level is a tough task.

**Reply by BurntSushi** [reply]
For streaming, @pascalkuthe is working on that! https://docs.rs/regex-cursor/latest/regex_cursor/

**Reply by cosmicexplorer** [reply]
> are primarily throughput optimizations. They aren't things that are likely to help with "parsing" tasks

This is totally correct! Geoff Langdale also once described traditional "lexing" to me as extremely difficult to optimize because you essentially have to restart the matching after every token (this is related to the implicit assumption that token spans do not overlap). I'm taking a very abstract interpretation of "parsing" and my specific goal is to leverage throughput optimizations so that more complex types of string matching logic can be achieved in a safe and correct way while maintaining throughput!

We've previously discussed why the hyperscan callback API is interesting, and I wrote a [small treatise](https://docs.rs/vectorscan-async/latest/vectorscan/matchers/index.html#advantages) in my docs for my rust wrapper lib on how the user callback is one (imperfect) way that hyperscan achieves throughput + user-defined match logic. I actually had been working on such a parser composition API for a while before ever looking into hyperscan, but didn't know of a good use case for it before I saw hyperscan's example recently and wrapped its API so I could understand its flaws.

**Reply by cosmicexplorer** [reply]
> The SIMD is all about quickly finding *the next candidate start position of a match*

So the other reason I wanted to try this idea is because I realized that SIMD prefiltering e.g. for literal substrings is actually separate from most other regex logic, and I wanted to see if I could architect a regex engine with fast prefiltering but decoupled from the actual literal matching. I then also realized that "finding candidate match boundaries" is actually very similar to a particular mechanic I've already developed separately to parse non-contiguous data! So as a hypothesis I'm seeing if "a protocol for communicating half-match boundaries" is a genuinely useful organizing principle to efficiently implement prefiltering mechanisms at multiple scales of linguistic complexity/abstraction.
